damnit/gui/zulip_messenger.py

```diff
@@ -0,0 +1,304 @@
+import zulip
+import logging
+import traceback
+import pandas as pd
+
+from configparser import ConfigParser
+from pathlib import Path
+from PyQt5 import QtWidgets, QtGui, QtCore
+
+log = logging.getLogger(__name__)
+
+# This class should be instantiated only per opened GUI. It's only propose is to
+# to hold cache information as well as a Zulip client, which might be updated.
+class ZulipMessenger():
+    def __init__(self, parent = None):
+        self.main_window = parent
+        self.config_path = Path.home() / ".local" / "state" / "damnit" / ".zuliprc"
+        self.config_path.parent.mkdir(parents=True, exist_ok=True)
+        self.key, self.email, self.stream, self.topic = '','','',''
+        self.client = None
+        self.streams = []
+        self.topics = []
+    
+    def send_table(self, tb):
+        config_dialog = ZulipConfig(self.main_window, self, tb, kind = 'table')
+        config_dialog.exec()
+        
+    def send_figure(self, fn):
+        config_dialog =  ZulipConfig(self.main_window, self, fn, kind='figure')
+        config_dialog.exec()
+        
+# This class handles the zulip request within a QDialog. One instance is created
+# per right click action (from e.g. the table view or the plot canvas) 
+class ZulipConfig(QtWidgets.QDialog):    
+    def __init__(self, parent = None, messenger = None, msg = None, kind = None):
+        super().__init__(parent)        
+        self.main_window = parent
+        self.messenger = messenger
+        self.resize(600, 300)
+        self.setWindowTitle("Zulip configuration")
+        self.setModal(False)
+        self.config_path = self.messenger.config_path
+        self.msg = msg
+        #E.g. table or figure
+        self.kind = kind
+
+        layout = QtWidgets.QGridLayout()        
+
+        self.setLayout(layout)
+        self._set_layout(layout)
+        
+    def enable_config(self, logic = True):
+        self.edit_email.setEnabled(logic)
+        self.edit_key.setEnabled(logic)
+        if not logic:
+            self.edit_key.setEchoMode(2)
+        else:
+            self.edit_key.setEchoMode(0)
+        
+    def search_streams(self):
+        self.show_msg('Fetching streams and topics, please wait', level = 'warning')
+        timer = QtCore.QTimer(self)
+        timer.setSingleShot(True)
+        timer.timeout.connect(self._search_streams)
+        timer.start()
+        
+    def _search_streams(self):
+        changes =  self.check_for_changes(include_streams=False)
+        if changes:
+            self.update_client()
+            
+        try:
+            streams = self.messenger.client.get_streams()
```

Comment:
it's probably faster to use [get_subscriptions](https://zulip.com/api/get-subscriptions), you should then filter based on the current proposal open by DAMNIT.